README.md
# JSON Schema Validator
[![Build Status](https://travis-ci.org/fpenim/json-schema-validator.svg?branch=dev)](https://travis-ci.org/fpenim/json-schema-validator) [![Codacy Badge](https://api.codacy.com/project/badge/Grade/7fbabc981e294249a9a0967965418058)](https://www.codacy.com/app/fpenim/json-schema-validator?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=EMBL-EBI-SUBS/json-schema-validator&amp;utm_campaign=Badge_Grade)
[![tested with jest](https://img.shields.io/badge/tested_with-jest-99424f.svg)](https://github.com/facebook/jest)

This repository contains a [JSON Schema](http://json-schema.org/). This validator runs as a standalone node server that receives validation requests and gives back it's results.
The validation is done using the [AJV](https://github.com/epoberezkin/ajv) library version ^6.0.0 that fully supports the JSON Schema **draft-07**.

Latest released version available [here](http://json-schema-validator.fpenim.com).

## Contents
  - [Getting Started](README.md#getting-started)

  - [Validation API](README.md#validation-api)

  - [Custom keywords](README.md#custom-keywords)

  - [License](README.md#license)

## Getting Started
These instructions will get you a copy of the project up and running on your local machine for development and testing purposes.

### Prerequisites
To be able to run this project you'll need to have [Node.js](https://nodejs.org/en/about/) and [npm](https://www.npmjs.com/) installed in your machine.
npm is distributed with Node.js which means that when you download Node.js, you automatically get npm installed on your computer.

### Installing

#### Node.js / npm
  - Get Node.js: https://nodejs.org/en/ (v8.11.1 LTS)

  - If you use [Homebrew](https://brew.sh/) you can install node by doing: `brew install node`

After installation check that everything is correctly installed and which versions you are running:
```
node -v
npm -v
```

#### Project
Clone project and install dependencies:
```
git clone https://github.com/fpenim/json-schema-validator.git
cd json-schema-validator
npm install
```

### Running the Tests
```
npm test
```

### Executing
```
node src/server
```
The node server will run on port **3020** and will expose one endpoint: **/validate**.

#### Startup arguments

- logPath

If provided with a log path argument, the application will write the logs to a file on the specified directory with a 24h rotation. To provide the log path add a `logPath` property after the startup statement:
```
node src/server --logPath=/log/directory/path
```

- pidPath

If provided with a pid file path argument, the application will write the pid into the specified file. If no pid file argument is provided, the application will still create a pid file on the default path: `./server.pid`.
To provide the pid file path add a `pidPath` property after the startup statement:
```
node src/server --pidPath=/pid/file/path/server.pid
```
Note: This is the **file path** and not just the directory it will be written to.

### Executing with Docker
  1. Build docker image:
```
docker build -t subs/json-schema-validator .
```
  2. Run docker image:
```
docker run -p 3020:3020 -d subs/json-schema-validator
```

### Development
For development purposes using [nodemon](https://nodemon.io/) is useful. It reloads the application everytime something has changed on save time.
```
nodemon src/server
```

## Validation API
This validator exposes one endpoint that accepts `POST` requests.

### /validate
This endpoint expects the body to have the following structure:
```js
{
  "schemas": [],
  "entity": {},
  "rootSchemaId": ""
}
```
Where the schemas should contain at least one valid json schema to validate the entity against. Multiple schemas may be provided when these reference each other using the keyword `"$ref"`. When this is the case, the `"$id"` of the primary/root schema must also be provided.

**Single schema example:**
```js
{
  "schemas": [
    {
      "type": "object",
      "properties": {
        "foo": { "type": "string" },
        "bar": { "type": "integer" },
        "zed": { "type": "boolean" }
      }
    }
  ],
  "entity": {
    "foo": "abc",
    "bar": 123,
    "zed": true
  }
}
```

**Multiple schemas example:** 
```js
{
  "schemas": [
    {
      "$id": "http://example.com/schemas/schema.json",
      "type": "object",
      "properties": {
        "foo": { "$ref": "defs.json#/definitions/int" },
        "bar": { "$ref": "definitions.json#/definitions/str" },
        "abc": { "$ref": "defs.json#/definitions/array" }
      },
      "required": ["foo"]
    },
    {
      "$id": "http://example.com/schemas/defs.json",
      "definitions": {
        "int": { "type": "integer" },
        "array": { "$ref": "definitions.json#/definitions/nextarray" }
      }
    },
    {
      "$id": "http://example.com/schemas/definitions.json",
      "definitions": {
        "str": { "type": "string" },
        "nextarray": { "type": "string" }
      }
    }
  ],
  "rootSchemaId": "http://example.com/schemas/schema.json",
  "entity": {
    "foo": 3,
    "abc": ""
  }
}
```

### Output

Response with no validation errors:

HTTP status code `200`
```js
[]
```
An example of a validation response with errors:

HTTP status code `200`
```js
[
  {
    "errors": [
        "should have required property 'value'"
    ],
    "dataPath": ".attributes['age'][0].value"
  },
  {
    "errors": [
        "should NOT be shorter than 1 characters",
        "should match format \"uri\""
    ],
    "dataPath": ".attributes['breed'][0].terms[0].url"
  }
]
```
Where *errors* is an array of error messages for a given input identified by its path on *dataPath*. There may be one or more error objects within the response array. An empty array represents a valid validation result.

### API Errors
Sending malformed JSON or a body with either the schema or the submittable missing will result in an API error (the request will not reach the validation). 

- When sending malformed JSON:

  HTTP status code `400` - Bad Request
  ```js
  {
    "errors": "Malformed JSON please check your request body."
  }
  ```
- When any of the required body values is missing:

  HTTP status code `422` - Unprocessable Entity
  ```js
  {
    "errors": {
      "schema": {
        "location": "body",
        "param": "schema",
        "msg": "Required."
      },
      "object": {
        "location": "body",
        "param": "object",
        "msg": "Required."
      }
    }
  }
  ```

## Custom keywords
The AJV library supports the implementation of custom json schema keywords to address validation scenarios that go beyond what json schema can handle.

Two example custom keywords implementations are available under `examples/custom-keywords`, these are `isChildTermOf` and `isValidTerm`. For more details see the documentation on [CUSTOM KEYWORDS](CUSTOM_KW.md).

## License
 For more details about licensing see the [LICENSE](LICENSE.md).


CHANGELOG.md
# JSON Schema Validator Changelog

## Release 0.1.0
Initial release.

## Release vNext

### Bug fixes
  * Validator now returns an useful error message if a JSON Schema version other than _draft-07_ is specified. 


test/validator.test.js
const fs = require("fs");
const validate = require("../src/validation/validator");

test("Empty Schema (empty object)", () => {

  let errors = validate([{}], {});
  expect(errors).toBeNull();
});

test("Attributes Schema", () => {
  let inputSchema = fs.readFileSync("examples/schemas/attributes-schema.json");
  let jsonSchema = JSON.parse(inputSchema);

  let inputObj = fs.readFileSync("examples/entities/attributes.json"); 
  let jsonObj = JSON.parse(inputObj);

  let errors = validate([jsonSchema], jsonObj);

  expect(errors).toBeDefined();
  expect(errors.length).toBe(1);

  expect(errors[0].dataPath).toBe(".attributes['breed'][0].terms[0].url");
  expect(errors[0].message).toBe("should match format \"uri\"");
});

test("BioSamples Schema - FAANG \'organism\' sample", () => {
  let inputSchema = fs.readFileSync("examples/schemas/biosamples-schema.json");
  let jsonSchema = JSON.parse(inputSchema);

  let inputObj = fs.readFileSync("examples/entities/faang-organism-sample.json");
  let jsonObj = JSON.parse(inputObj);

  let errors = validate([jsonSchema], jsonObj);
  expect(errors).toBeNull();
});

test("Study Schema", () => {
  let inputSchema = fs.readFileSync("examples/schemas/submittables/study-schema.json");
  let jsonSchema = JSON.parse(inputSchema);

  let inputObj = fs.readFileSync("examples/entities/study.json");
  let jsonObj = JSON.parse(inputObj);

  let errors = validate([jsonSchema], jsonObj);
 
  expect(errors).toBeDefined();
  expect(errors.length).toBe(2);
 
  expect(errors[0].dataPath).toBe("");
  expect(errors[0].message).toBe("should have required property 'alias'");
  expect(errors[1].dataPath).toBe("");
  expect(errors[1].message).toBe("should have required property 'StudyDataType'");
});

test/schema-references.test.js
const fs = require("fs");
const validate = require("../src/validation/validator");

test("Base schema and definitions schema test WITH erros", () => {
    let baseSchema = fs.readFileSync("examples/schemas/references/base-sample-schema.json");
    let jsonBaseSchema = JSON.parse(baseSchema);

    let definitionsSchema = fs.readFileSync("examples/schemas/references/definitions-schema.json");
    let jsonDefinitionsSchema = JSON.parse(definitionsSchema);

    let errors = validate(
        [jsonBaseSchema, jsonDefinitionsSchema], 
        {
            alias: "abc",
            taxonId: 9606,
            releaseDate: "2000-01-01",
            sampleRelationships: [{
                accession: "def123",
                relationshipNature: ""
            }]
        },
        jsonBaseSchema.$id
    );

    expect(errors).toBeDefined();
    expect(errors[0].schemaPath).toBe("definitions-schema.json#/definitions/sampleRelationships/items/properties/relationshipNature/enum");
});

test("Base schema and definitions schema test WITHOUT errors", () => {
    let baseSchema = fs.readFileSync("examples/schemas/references/base-sample-schema.json");
    let jsonBaseSchema = JSON.parse(baseSchema);

    let definitionsSchema = fs.readFileSync("examples/schemas/references/definitions-schema.json");
    let jsonDefinitionsSchema = JSON.parse(definitionsSchema);

    let errors = validate(
        [jsonBaseSchema, jsonDefinitionsSchema],
        {
            alias: "abc",
            taxonId: 9606,
            releaseDate: "2000-01-01",
            sampleRelationships: [{
                accession: "def123",
                relationshipNature: "derived from"
            }]
        },
        jsonBaseSchema.$id
    );

    expect(errors).toBeNull();
});

test("ML schema -> base schema -> definitions schema test", () => {
    let mlSchema = fs.readFileSync("examples/schemas/references/ml-sample-schema.json");
    let jsonMLSchema = JSON.parse(mlSchema);

    let baseSchema = fs.readFileSync("examples/schemas/references/base-sample-schema.json");
    let jsonBaseSchema = JSON.parse(baseSchema);

    let definitionsSchema = fs.readFileSync("examples/schemas/references/definitions-schema.json");
    let jsonDefinitionsSchema = JSON.parse(definitionsSchema);

    let errors = validate(
        [jsonMLSchema, jsonBaseSchema, jsonDefinitionsSchema], 
        {
            alias: "abc",
            taxonId: 9606,
            releaseDate: "2000-01-01",
            attributes: {
                Organism: [{value: "Homo sapiens"}],
                "Organism part": [{}]
            }
        },
        jsonMLSchema.$id
    );

    expect(errors[0].schemaPath).toBe("definitions-schema.json#/definitions/attributes_structure/items/required");
});

src/server.js
const express = require("express");
const logger = require("./winston");
const AppError = require("./model/application-error");
const { check, validationResult } = require("express-validator/check");
const { handleValidation } = require("./validation/validation-handler");

const argv = require("yargs").argv;
const npid = require("npid");

const app = express();
const port = process.env.PORT || 3020;

app.use(express.json());

app.use(function(req, res, next) {
  res.header("Access-Control-Allow-Origin", "*");
  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
  next();
});

app.use(function(err, req, res, next) {
  if (err instanceof SyntaxError && err.status === 400 && "body" in err) {
    let appError = new AppError("Received malformed JSON.");
    logger.log("info", appError.errors);
    res.status(400).send(appError);
  } else {
    let appError = new AppError(err.message);
    logger.log("error", appError.errors);
    res.status(err.status).send(appError);
  }
});

// -- Endpoint definition -- //
app.post("/validate", [
    check("schemas", "At least one schema to validate the entity against is required.").isArray().not().isEmpty(),
    check("entity", "An entity to be validated is required.").exists()
  ], (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(422).json({ errors: errors.mapped() });
    } else {
      logger.log("debug", "Received POST request.");
      try {
        logger.log("debug", "Triggered validation . . .");
        let errors = handleValidation(req.body.schemas, req.body.entity, req.body.rootSchemaId);
        return res.json(errors || []);
      } catch(err) {
        logger.log("error", err.stack);
        return res.status(500).send(new AppError("An unexpected error occured: " + err.message));
      }
    }
  }
);

app.get("/validate", (req, res) => {
  logger.log("silly", "Received GET request.");
  res.send({
    message: "This is a JSON Schema Validator. Please POST to this endpoint the schema and entity to validate structured as showed bellow.",
    body: {
      schemas: [{}],
      entity: {}
    },
    repository: "https://github.com/fpenim/json-schema-validator"
  });
});

app.get("/", (req, res) => {
  res.redirect("/validate");
});

app.listen(port, () => {
  logger.log("info", ` -- Started server on port ${port} --`);
  if(argv.logPath) { logger.log("info", ` --> Log output: ${argv.logPath}`); }
});

// -- For monitoring purposes -- //
const pidPath = argv.pidPath || "./server.pid";
try {
  let pid = npid.create(pidPath);
  pid.removeOnExit();
} catch(err) {
  logger.log("error", err);
  process.exit(1);
}

// Handles crt + c event
process.on("SIGINT", () => {
  npid.remove(pidPath);
  process.exit();
});

// Handles kill -USR1 pid event (monit)
process.on("SIGUSR1", () => {
  npid.remove(pidPath);
  process.exit();
});

//Handles kill -USR2 pid event (nodemon)
process.on("SIGUSR2", () => {
  npid.remove(pidPath);
  process.exit();
});